### DISCUSSION TITLE
Python dependencies

### DISCUSSION THREAD

**Discussion by azhuchkov** [Homebrew/discussions#4919] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.20-23-g25d1f10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 25d1f1064aa2f79e1a01d7418b920b85a99c1f03
Last commit: 4 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 464711680f803810cbb976fb0e8a7b842eed38b2
Core tap last commit: 7 days ago
Core tap branch: master
Core tap JSON: 16 Nov 19:20 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BAT_THEME: DarkNeon
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.VemhkMdZcL/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_ENV_HINTS: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 14.0.0 build 1400
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 12.7.1-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  node@10
  unrar

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  libbpg
  node@10
  openssl@1.1
  sdl_image

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  yarn
  kubernetes-cli

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
```

### Description of issue

I'm trying to build a formula which includes Python script with dependencies. I create a virtualenv during installation and install every dependency into it. For that I did something like:
```ruby
...
venv = virtualenv_create(libexec, "python3")
venv.pip_install "jq==1.6.0"
...
```

And it actually works! The problem I see is that `pip_install` does not install transitive dependencies (uses `--no-deps` option). Also installation is quite slow due to building from the sources.

So, I'd like to know what is the purposes of it if I do everything correctly. And how to fix that if it's unintended behaviour for those cases like mine.

Thanks!

**Comment by p-linnane** [comment]
Please see [`abi3audit`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/a/abi3audit.rb) as an example of how to do what you are trying to do. You can run `brew update-python-resources` followed by your formula name to populate the resources.

**Reply by azhuchkov** [reply]
Thanks for the reply! I appreciate your help.

I suppose that example also builds all the resources from scratch (instead of usual PIP installation), but I want to find some explanation about it, I mean why it works the way it works.

Also `brew update-python-resources` command hanging in my case. I suppose the reason of it is I don't have PyPi package for my formula.

**Reply by p-linnane** [reply, marked as answer]
Everything in `homebrew-core` builds from source. And yes, not having a PyPI package will cause issues with `brew update-python-resources`.

**Comment by alexmogan0123** [comment]
I'm trying to build a formula which includes Python script with dependencies. I create a virtualenv during installation and install every dependency into it. For that I did something like:

...
venv = virtualenv_create(libexec, "python3")
venv.pip_install "jq==1.6.0"
...
And it actually works! The problem I see is that pip_install does not install transitive dependencies (uses --no-deps option). Also installation is quite slow due to building from the sources.

So, I'd like to know what is the purposes of it if I do everything correctly. And how to fix that if it's unintended behaviour for those cases like mine.

Thanks!

**Comment by mianyangnu** [comment]
I'm trying to build a formula which includes Python script with dependencies. I create a virtualenv during installation and install every dependency into it. For that I did something like:

...
venv = virtualenv_create(libexec, "python3")
venv.pip_install "jq==1.6.0"
...
And it actually works! The problem I see is that pip_install does not install transitive dependencies (uses --no-deps option). Also installation is quite slow due to building from the sources.

So, I'd like to know what is the purposes of it if I do everything correctly. And how to fix that if it's unintended behaviour for those cases like mine.

Thanks!
